code/modules/unit_tests/barsigns.dm

```diff
@@ -0,0 +1,41 @@
+/**
+ * Test if icon states for each datum actually exist in the DMI.
+ */
+/datum/unit_test/barsigns_icon/Run()
+	var/obj/structure/sign/barsign_type = /obj/structure/sign/barsign
+	var/icon/barsign_icon = initial(barsign_type.icon)
+	var/list/barsign_icon_states = icon_states(barsign_icon)
+
+	// Check every datum
+	for(var/sign_type in subtypesof(/datum/barsign))
+		// Special supertype
+		if(sign_type == /datum/barsign/hiddensigns)
+			continue
+
+		var/datum/barsign/sign = new sign_type()
+
+		if(!(sign.icon in barsign_icon_states))
+			TEST_FAIL("Icon state for [sign_type] does not exist in [barsign_icon].")
+
+/**
+ * Check that bar signs have a name and desc, and that the name is unique.
+ */
+/datum/unit_test/barsigns_name/Run()
+	var/list/existing_names = list()
+
+	for(var/sign_type in subtypesof(/datum/barsign))
+		var/datum/barsign/sign = new sign_type()
+
+		// Hidden bar signs aren't expected to have a name.
+		if(sign.hidden)
```

Comment:
Why do you check `.hidden` here, but an explicit typepath above?

Reply:
I mean, I could do `subtypesof(/datum/barsign) - typesof(/datum/barsign/hidden)` but in theory there could be `hidden=TRUE` subtypes anywhere in the tree.

Reply:
So why don't they *both* check hidden?

Reply:
The subtypes of `/datum/barsign/hidden` should also have valid icon states, as they are used, just for special circumstances. Just specifically `/datum/barsign/hidden` has no icon state as it is an abstract supertype.

However, the empbarsign and signoff ones have explicitly null names (I have no clue why, but I'm sure there's some esoteric reason, but they are hidden and don't rename the room, so who knows) and thus would fail the name check. If you would like, I can give them names so that both checks can just do `subtypesof(/datum/barsign) - typesof(/datum/barsign/hidden)`

Reply:
Giving them all names is probably a good idea regardless

Reply:
They now have names except the `/datum/barsign` and `/datum/barsign/hidden` baseclasses which should crash if selected anyway as something has horribly gone wrong if they do.

I cleaned up some of the code that checked for names existing as that is now unnecessary. Both unit tests use `subtypesof(/datum/barsign) - typesof(/datum/barsign/hidden)` now.